src/borrow.rs

```diff
@@ -0,0 +1,168 @@
+use std::cell::UnsafeCell;
+use std::collections::hash_map::{Entry, HashMap};
+use std::ops::{Deref, DerefMut};
+
+use ndarray::{ArrayView, ArrayViewMut, Dimension};
+use pyo3::{FromPyObject, PyAny, PyResult};
+
+use crate::array::PyArray;
+use crate::dtype::Element;
+
+thread_local! {
+    static BORROW_FLAGS: UnsafeCell<HashMap<usize, isize>> = UnsafeCell::new(HashMap::new());
+}
+
+pub struct PyArrayRef<'a, T, D> {
+    array: &'a PyArray<T, D>,
+    view: ArrayView<'a, T, D>,
+}
+
+impl<'a, T, D> Deref for PyArrayRef<'a, T, D> {
+    type Target = ArrayView<'a, T, D>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.view
+    }
+}
+
+impl<'py, T: Element, D: Dimension> FromPyObject<'py> for PyArrayRef<'py, T, D> {
+    fn extract(obj: &'py PyAny) -> PyResult<Self> {
+        let array: &'py PyArray<T, D> = obj.extract()?;
+        Ok(array.as_array())
+    }
+}
+
+impl<'a, T, D> PyArrayRef<'a, T, D>
+where
+    T: Element,
+    D: Dimension,
+{
+    pub(crate) fn try_new(array: &'a PyArray<T, D>) -> Option<Self> {
+        let address = array as *const PyArray<T, D> as usize;
```

Comment:
As written in #258, it might actually make sense to follow the "base object chain" here and use the address of the final base object as the key into the hash table which would be able to safely detect quite a few aliasing issues on the Python side of the fence as well.

But it would also be an over-approximation that could sometimes fail unexpectedly if it does not consider e.g. splitting a large array into multiple non-overlapping slices which can then be borrowed separately.

Reply:
I would really like to find out if there are projects out there which break if the over-approximation, i.e. distinct borrows must have distinct root base objects, is taken. Of course, I would like to find out without releasing that code into the wild and waiting for the angry complaints. :-(

Reply:
(There is of course another issue where this might not work properly, considering that base objects are not necessarily NumPy arrays themselves (e.g. our `PySliceContainer`), there is probably no real guarantee that distinct base objects mean non-overlapping backing memory, e.g. memory map base objects could be distinct while pointing into the same address range.)

Reply:
Do you think it's practical for us to calculate the start and end addresses of the memory pointed at by the view? If we used a binary-search based structure to store all the these (start, end) tuples we may be able to detect overlaps reasonably efficiently.

(I would hope that the cost of doing these aliasing checks will work out as much cheaper than large array operations!)

Reply:
Is this sufficient though? Isn't possible that those ranges overlap even for non-overlapping array views due to non-unit strides? For example, have a large arrays with three RGB layers which gets sliced into three separate views where all have basically the same start-to-end range but never overlap?

Reply:
Ah, right. I would hope that non-unit strides are the exception (so might be able to optimise for the common case). They may turn out to be very common though, e.g. 1-D slices of multidimensional arrays.

Reply:
I fear that the general case may be an instance of satisfiability and therefore NP: For two arrays `A_1/2` with the same base object and data pointers `p_1/2` and dimensionalities `D_1,2` and dimensions `d_{1/2,i}` and strides `s_{1/2,i}`, we need to determine whether the equation
```
p_1 + s_1,1 * k_1 + ... + s_{1,D_1} k_{D_1} = p_2 +  s_2,1 * l_1 + ... + s_{2,D_2} * l_{D_2}
```
has integer solutions in the domain
```
k_i in {0, ..., d_{1,i} - 1} for i in {1, ..., D_1}
l_i in {0, ..., d_{2,i} - 1} for i in {1, ..., D_2}
```

Not sure whether there is any structure in this problem which makes this substantially easier...